### ISSUE TITLE
DDP掉精度

### ISSUE THREAD

**Issue by lx-kika** [tczhangzhi/pytorch-distributed#3]
您好，我使用的是DistributedDataParallel，通信是nccl，然后数据方面使用的DistributedSampler，但是发现对比单卡同样参数设置的模型，多卡的精度会大幅下降（模型里也没有BN层）。
然后还有几个疑问望解答：
1. 多卡相当于增大batch size吗？
2.假如使用warm up的话，对应设置和单卡时应该除以卡的数量吗？
3.使用adamw自适应优化器，多卡时需要将lr乘以对应卡数吗？
谢谢

**Comment by laoreja** [comment]
参考一下 linear scale learning rate？
根据batch size （汇总所有机器gpu后的batch size） linear scale learning rate。

**Comment by lx-kika** [comment]
嗯这个我也试过，分别按batch size线性和开方两种比例调整lr但是最后结果还是不一样。

**Comment by tczhangzhi** [comment]
抱歉最近没怎么看 issue，这次 github 更新了 UI 在 timeline 上看到了。来补充一下希望可以帮到。

使用示例中的 DistributedDataParallel 的话，可以理解为 batch size 不变，即在命令行中传入的 batch size 和 lr 应该与 single gpu 对应 (例如，等比)。在实现过程中，我们把 batch size 个数据切分成 ngpus_per_node，分块放入不同 GPU，在计算梯度后对这 ngpus_per_node 块梯度进行平均再重新分发梯度，更新所有 GPU 上的参数。

可以这样简单理解：在图像分类中，如果我们将 single gpu 上 loss 的计算过程（实际应该是梯度的计算过程，这里简化了）看作对 batch 中每个图像分别计算交叉熵，求这些交叉熵的平均用于梯度更新。那么 multi gpu 上 loss 的计算过程就是将 batch 中的图像分成 ngpus_per_node 块，每一块求其中所有图像的交叉熵的平均，再对 ngpus_per_node 块求平均用于梯度更新。由于每块中的图像数量相同，其结果应当是一致的。

当然，在实际操作过程中，无法将问题这样简化因此会产生些许误差。不过目前在我的实践中误差仍可以接受。另外我也有在单机的 NAS 过程中遇到 batch size 增大，最优 lr 不成线性增长的现象的。因此影响因素还是比较多样的。
